internal/dag/builder.go

```diff
@@ -917,29 +919,94 @@ func (b *Builder) processTCPProxy(sw *ObjectStatusWriter, ir *ingressroutev1.Ing
 
 		// follow the link and process the target ingress route
 		sw, commit := sw.WithObject(dest)
-		b.processTCPProxy(sw, dest, visited, host)
+		b.processIngressRouteTCPProxy(sw, dest, visited, host)
 		commit()
 	}
 }
 
-func (b *Builder) processTCPProxyHTTPProxy(sw *ObjectStatusWriter, httpproxy *projcontour.HTTPProxy, host string) {
-	if len(httpproxy.Spec.TCPProxy.Services) > 0 {
+// processHTTPProxyTCPProxy processes the spec.tcpproxy stanza in a HTTPProxy document
+// following the chain of spec.tcpproxy.include references. It returns true if processing
+// was successful, otherwise false if an error was encountered. The details of the error
+// will be recorded on the status of the relevant HTTPProxy object,
+func (b *Builder) processHTTPProxyTCPProxy(sw *ObjectStatusWriter, httpproxy *projcontour.HTTPProxy, visited []*projcontour.HTTPProxy, host string) bool {
+	tcpproxy := httpproxy.Spec.TCPProxy
+	if tcpproxy == nil {
+		// nothing to do
+		return true
+	}
+
+	visited = append(visited, httpproxy)
+
+	if len(tcpproxy.Services) > 0 && tcpproxy.Include != nil {
+		sw.SetInvalid("tcpproxy: cannot specify services and include in the same httpproxy")
+		return false
+	}
+
```

Comment:
Should we add a check that one of services or include should be set?

Reply:
Good catch. There is a check that one of services or include are set, but nothing happens if neither are set. This could happen directly
```
spec:
  tcpproxy:
```
or transitively
spec:
  tcpproxy:
     include:
       name: bar
       namespace: baz
---
#baz/bar
spec: 
   tcpproxy:
```
I've raised #1776 to track this.